File: built-in plugins dsp/mb-time-retrigger.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbTimeRetriggerPlugin: PluginDefinition = { id: 'mb-time-retrigger', slug: 'mb-time-retrigger', name: 'MB Re-Trigger', category: 'effect', type: 'delay' as any, version: '1.0.0', description: 'Capture audio buffer and retrigger with pitch ramp', author: 'Max Booster', parameters: [{ id: 'bufferSize', name: 'Buffer Size', type: 'float', defaultValue: 100, minValue: 10, maxValue: 1000, automatable: true }, { id: 'rate', name: 'Rate', type: 'float', defaultValue: 4, minValue: 1, maxValue: 32, automatable: true }, { id: 'pitchRamp', name: 'Pitch Ramp', type: 'float', defaultValue: 0, minValue: -12, maxValue: 12, automatable: true }, { id: 'mix', name: 'Mix', type: 'float', defaultValue: 1, minValue: 0, maxValue: 1, automatable: true }], defaultPreset: { bufferSize: 100, rate: 4, pitchRamp: 0, mix: 1 } };

export default MbTimeRetriggerPlugin;

File: built-in plugins dsp/mb-time-retrigger.cpp
/**
 * MB Re-Trigger
 * Category : effect
 * Type     : delay
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Capture audio buffer and retrigger with pitch ramp
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_TIME_RETRIGGER_H
#define MB_TIME_RETRIGGER_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbTimeRetrigger : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-time-retrigger";
    static constexpr const char* PLUGIN_NAME    = "MB Re-Trigger";
    static constexpr const char* PLUGIN_TYPE    = "delay";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float bufferSize = 100f;  // range [10, 1000]
    float rate = 4f;  // range [1, 32]
    float pitchRamp = 0f;  // range [-12, 12]
    float mix = 1f;  // range [0, 1]
    };

    MbTimeRetrigger() = default;
    ~MbTimeRetrigger() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.bufferSize = std::clamp(params.bufferSize, 10f, 1000f);
        params.rate = std::clamp(params.rate, 1f, 32f);
        params.pitchRamp = std::clamp(params.pitchRamp, -12f, 12f);
        params.mix = std::clamp(params.mix, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Re-Trigger
        return input;
    }
};

#endif // MB_TIME_RETRIGGER_H
